.. pyranges documentation master file, created by
   sphinx-quickstart.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.


The pyranges documentation
==========================
Pyranges is a Python library specifically designed for efficient and intuitive manipulation of genomics data,
particularly genomic intervals (like genes, genomic features, or reads).
The library is optimized for fast querying and manipulation of genomic annotations.

Pyranges is open source, and hosted at github: https://github.com/pyranges/pyranges_1.x/

Pyranges is developed by Endre Bakken Stovner and by
`Marco Mariotti's lab <https://www.mariottigenomicslab.com/>`_.

**This documentation refers to the version 1 of pyranges**, which introduces a new API and a new data structure
compared to version 0, still the 'default', soon to be deprecated. If you are a pyranges v0 user,
check :doc:`this guide to migrate to v1 <./migration_guide>`.


Citation
~~~~~~~~

Stovner EB, SÃ¦trom P (2020) PyRanges: efficient comparison of genomic intervals in Python. *Bioinformatics 36(3):918-919*  http://dx.doi.org/10.1093/bioinformatics/btz615


Documentation outline
~~~~~~~~~~~~~~~~~~~~~

#. ðŸš€ :doc:`Installation instructions <./installation>`
#. ðŸš€ :doc:`The tutorial <./tutorial>`,  recommended for all new users
#. ðŸš€ :doc:`The how-to pages <./how_to_pages>`, further below, where functionalities are grouped by topic
#. ðŸš€ :doc:`The API reference <./api_reference>`, where all methods are explained in detail
#. ðŸš€ :doc:`The developer guide <./developer_guide>`, to follow in order to contribute to PyRanges
#. ðŸš€ :doc:`The guide to migrate to v1 <./migration_guide>`, for existing users of pyranges v0



.. toctree::
   :maxdepth: 1
   :hidden:
   :caption: Contents:

   installation
   tutorial
   how_to_pages
   api_reference
   developer_guide
   migration_guide


Indices
=======

* :ref:`genindex`
* :ref:`modindex`
